src/model_io/codecs/src/URDFDocument.cpp

```diff
@@ -6,16 +6,278 @@
 
 #include <iDynTree/PrismaticJoint.h>
 #include <iDynTree/RevoluteJoint.h>
+#include <iDynTree/SphericalJoint.h>
+#include <iDynTree/FixedJoint.h>
 #include <iDynTree/ModelTransformers.h>
 #include <iDynTree/SixAxisForceTorqueSensor.h>
 
 #include <algorithm>
 #include <unordered_set>
 #include <vector>
+#include <set>
 
 namespace iDynTree {
 
-    // Functions local to the current compilation unit
+        struct ThreeRevoluteJointPattern {
+        std::string parentLinkName;
+        std::string intermediateLinkName1;
+        std::string intermediateLinkName2;
+        std::string childLinkName;
+
+        std::string revJointName1; // X-axis rotation
+        std::string revJointName2; // Y-axis rotation
+        std::string revJointName3; // Z-axis rotation
+
+        Transform parentLink_H_childLink;
+        Position jointCenter; // In parent link frame
+    };
+
+    // Helper function to check if three vectors are orthogonal
+    bool areVectorsOrthogonal(const Direction& v1, const Direction& v2, const Direction& v3, double tolerance) {
+        return v1.isPerpendicular(v2, tolerance) &&
+               v1.isPerpendicular(v3, tolerance) &&
+               v2.isPerpendicular(v3, tolerance);
+    }
+
+    // Helper function to find intersection point of three lines
+    bool findAxisIntersectionPoint(const Axis& axis1, const Axis& axis2, const Axis& axis3,
+                                   Position& intersectionPoint, double tolerance) {
+        // For simplicity, we'll check if all axes pass through the same point
+        // by checking if the distance from each axis to a common point is within tolerance
+
+        // Use axis1's origin as reference point
+        Position refPoint = axis1.getOrigin();
+
+        double dist2 = axis2.getDistanceBetweenAxisAndPoint(refPoint);
+        double dist3 = axis3.getDistanceBetweenAxisAndPoint(refPoint);
+
+        if (dist2 < tolerance && dist3 < tolerance) {
+            intersectionPoint = refPoint;
+            return true;
+        }
+
+        return false;
+    }
+
+    // Function to detect and convert three consecutive revolute joints to spherical joint
+    bool convertThreeRevoluteJointsToSpherical(Model& model,
```

Comment:
Not fundamental, but all others functions that process models somehow (like that a `iDynTree::Model` enters and another modified `iDynTree::Model` exists) are contained in https://github.com/robotology/idyntree/blob/23f139731e443ef4215c01b3c3204856c903bc8d/src/model/include/iDynTree/ModelTransformers.h#L5 and https://github.com/robotology/idyntree/blob/23f139731e443ef4215c01b3c3204856c903bc8d/src/model/src/ModelTransformers.cpp, so that they can be tested and reused indipendently from URDF import and export. Could it make sense to do the same also for this functions?

Reply:
Absolutely!